presentation/backend/src/presentation-backend/ElementPropertiesHelper.ts

```diff
@@ -37,70 +34,285 @@ function getECSqlName(fullClassName: string) {
+      filter += " OR ";
     }
+    filter += cond;
   }
-  if (fromId && currId) {
-    batches.push({ from: fromId, to: currId });
+  for (const item of batch) {
+    if (item.from === item.to) {
+      appendCondition(`this.ECInstanceId = ${item.from}`);
+    } else {
+      appendCondition(`this.ECInstanceId >= ${item.from} AND this.ECInstanceId <= ${item.to}`);
+    }
   }
-  return batches;
+  return filter;
 }
 
-function createElementsFilter(elementAlias: string, classNames?: string[]) {
-  if (classNames === undefined || classNames.length === 0) {
-    return undefined;
-  }
+function createClassContentRuleset(fullClassName: string): Ruleset {
+  const [schemaName, className] = parseFullClassName(fullClassName);
+  return {
+    id: `content/class-descriptor/${fullClassName}`,
+    rules: [
+      {
+        ruleType: "Content",
+        specifications: [
+          {
+            specType: "ContentInstancesOfSpecificClasses",
+            classes: {
+              schemaName,
+              classNames: [className],
+              arePolymorphic: false,
+            },
+            handlePropertiesPolymorphically: true,
+          },
+        ],
+      },
+    ],
+  };
+}
 
-  // check if list contains only valid class names
-  const classNameRegExp = new RegExp(/^[\w]+[.:][\w]+$/);
-  const invalidName = classNames.find((name) => !name.match(classNameRegExp));
-  if (invalidName) {
-    throw new PresentationError(
-      PresentationStatus.InvalidArgument,
-      `Encountered invalid class name - ${invalidName}.
-      Valid class name formats: "<schema name or alias>.<class name>", "<schema name or alias>:<class name>"`,
-    );
-  }
+/** Given a list of element ids, group them by class name. */
+function getElementClassesFromIds(imodel: IModelDb, elementIds: string[]): Observable<{ classFullName: string; ids: Id64Array }> {
+  const elementIdsBatchSize = 5000;
+  return range(0, elementIds.length / elementIdsBatchSize).pipe(
+    mergeMap((batchIndex) => {
+      const idsFrom = batchIndex * elementIdsBatchSize;
+      const idsTo = Math.min(idsFrom + elementIdsBatchSize, elementIds.length);
+      let idsFilter = "";
+      for (let i = idsFrom; i < idsTo; i++) {
+        idsFilter += `${elementIds[i]}`;
+        if (i < idsTo - 1) {
+          idsFilter += ",";
+        }
+      }
+      return from(
+        imodel.createQueryReader(
+          `
+            SELECT ec_classname(e.ECClassId) className, GROUP_CONCAT(IdToHex(e.ECInstanceId)) ids
+            FROM bis.Element e
+            WHERE e.ECInstanceId IN (${idsFilter})
+            GROUP BY e.ECClassId
+          `,
+        ),
+      );
+    }),
+    map((row: QueryRowProxy): { className: string; ids: Id64Array } => ({ className: row.className, ids: row.ids.split(",") })),
+    groupBy(({ className }) => className),
+    mergeMap((groups) =>
+      groups.pipe(
+        reduce<ObservedValueOf<typeof groups>, { classFullName: string; ids: Id64Array }>(
```

Comment:
Why is this `reduce` operator is needed? Can we just do `map((g) => ({ classFullName: g.className, ids: g.ids}))`?

Reply:
The `groupBy` operator returns an observable (class groups) of observables (pairs of class name and id array). We need to merge ids of the same class into a single array.